File: skills/onchain-app-interfaces/scripts/onchain_interfaces/metadata.py
"""Required ERC-7572 application metadata and bounded URI resolution."""

from __future__ import annotations

import base64
import json
import re
import urllib.parse
from typing import Any

from .external import canonical_origin, execute_https_request

MAX_METADATA_BYTES = 65_536
MAX_METADATA_URI_BYTES = 3 * MAX_METADATA_BYTES + 256
_URI = re.compile(r"^[a-zA-Z][a-zA-Z0-9+.-]*:\S+$")
_DATA = re.compile(
    r"^data:application/json(?:;utf8|;charset=utf-8)?(;base64)?,([\s\S]*)$",
    re.IGNORECASE,
)


def _reject_constant(value: str) -> None:
    raise ValueError("Invalid JSON constant")


def parse_contract_metadata(*, body: bytes) -> dict[str, Any]:
    try:
        if len(body) > MAX_METADATA_BYTES:
            raise ValueError("Metadata exceeds client size limit")
        value = json.loads(
            body.decode("utf-8", errors="strict"), parse_constant=_reject_constant
        )
        if not isinstance(value, dict):
            raise ValueError("Metadata must be an object")
        for field in ("name", "description"):
            if not isinstance(value.get(field), str) or not value[field].strip():
                raise ValueError("Missing nonempty metadata text")
        if "symbol" in value and not isinstance(value["symbol"], str):
            raise ValueError("Invalid symbol")
        for field in ("image", "banner_image", "featured_image", "external_link"):
            if field in value and (
                not isinstance(value[field], str)
                or _URI.fullmatch(value[field]) is None
            ):
                raise ValueError("Invalid metadata URI")
        if "collaborators" in value and (
            not isinstance(value["collaborators"], list)
            or any(not isinstance(item, str) for item in value["collaborators"])
        ):
            raise ValueError("Invalid collaborators")
        return value
    except (ValueError, UnicodeError, RecursionError) as error:
        raise ValueError(
            "Invalid application metadata: expected ERC-7572 JSON with nonempty name and description"
        ) from error


def _unquote(*, value: str) -> bytes:
    if re.search(r"%(?![0-9a-fA-F]{2})", value):
        raise ValueError("Invalid percent escape")
    return urllib.parse.unquote_to_bytes(value)


def _remote_url(*, uri: str, ipfs_gateway: str | None) -> str:
    if uri.startswith("https://"):
        return uri
    if not uri.startswith("ipfs://"):
        raise ValueError("Unsupported metadata URI scheme")
    if not ipfs_gateway:
        raise ValueError("IPFS requires a configured HTTPS gateway")
    gateway = canonical_origin(ipfs_gateway, require_origin_only=True)
    match = re.fullmatch(r"ipfs://([a-zA-Z0-9]+)((?:/[^?#]*)?)", uri)
    if match is None:
        raise ValueError("Invalid IPFS URI")
    segments = []
    for segment in match[2].split("/")[1:]:
        decoded = _unquote(value=segment).decode("utf-8", errors="strict")
        if decoded in (".", "..") or "/" in decoded or "\\" in decoded:
            raise ValueError("Invalid IPFS path")
        segments.append(urllib.parse.quote(decoded, safe="~!*'()-._"))
    suffix = "/" + "/".join(segments) if segments else ""
    return gateway + "/ipfs/" + match[1] + suffix


def resolve_contract_metadata(
    *,
    uri: str,
    allowed_origins: set[str],
    ipfs_gateway: str | None = None,
    max_header_bytes: int = 65_536,
    timeout: float = 30.0,
) -> dict[str, Any]:
    if not uri:
        raise ValueError("Invalid application metadata: contractURI() is empty")
    if (
        len(uri) > MAX_METADATA_URI_BYTES
        or len(uri.encode("utf-8")) > MAX_METADATA_URI_BYTES
    ):
        raise ValueError("Metadata exceeds client size limit")
    if uri.lower().startswith("data:"):
        try:
            match = _DATA.fullmatch(uri)
            if match is None:
                raise ValueError("Unsupported JSON data URI encoding")
            body = _unquote(value=match[2])
            if match[1]:
                encoded = body
                body = base64.b64decode(encoded, validate=True)
                if base64.b64encode(body) != encoded:
                    raise ValueError("Noncanonical Base64")
            return parse_contract_metadata(body=body)
        except (ValueError, UnicodeError) as error:
            raise ValueError("Invalid application metadata data URI") from error
    try:
        url = _remote_url(uri=uri, ipfs_gateway=ipfs_gateway)
        response = execute_https_request(
            request={
                "url": url,
                "method": "GET",
                "headers": [{"name": "Accept", "value": "application/json"}],
                "body": b"",
            },
            allowed_origins=allowed_origins,
            max_response_bytes=MAX_METADATA_BYTES,
            max_header_bytes=max_header_bytes,
            timeout=timeout,
        )
        if not 200 <= response["status"] < 300:
            raise ValueError("Metadata HTTP status %d" % response["status"])
    except (ValueError, RuntimeError, OSError) as error:
        raise RuntimeError("Metadata resolution failed: %s" % error) from error
    return parse_contract_metadata(body=response["body"])

File: src/client/metadata.ts
import {
  decodeFunctionResult,
  encodeFunctionData,
  hexToBytes,
  type Address,
} from "viem";
import { z } from "zod";

import { executeHttpRequest, validateRequest } from "./request.ts";
import type {
  AuthorizeRequest,
  EthCall,
  HttpFetch,
  HttpRequest,
} from "./types.ts";

export const maxMetadataBytes = 65_536;
export const maxMetadataUriBytes = 3 * maxMetadataBytes + 256;

export const contractMetadataAbi = [
  {
    type: "function",
    name: "contractURI",
    stateMutability: "view",
    inputs: [],
    outputs: [{ name: "uri", type: "string" }],
  },
  { type: "event", name: "ContractURIUpdated", inputs: [] },
] as const;

const nonemptyText = z.string().refine((value) => value.trim().length > 0);
const uriSchema = z.string().regex(/^[a-zA-Z][a-zA-Z0-9+.-]*:\S+$/);

export const contractMetadataSchema = z.looseObject({
  name: nonemptyText,
  description: nonemptyText,
  symbol: z.string().optional(),
  image: uriSchema.optional(),
  banner_image: uriSchema.optional(),
  featured_image: uriSchema.optional(),
  external_link: uriSchema.optional(),
  collaborators: z.array(z.string()).optional(),
});

export type ContractMetadata = z.infer<typeof contractMetadataSchema>;
export type ResolvedContractMetadata = {
  uri: string;
  metadata: ContractMetadata;
};
export type MetadataResolutionOptions = {
  authorizeRequest?: AuthorizeRequest;
  fetch?: HttpFetch;
  /** HTTPS gateway origin, e.g. https://ipfs.io. Its origin must be authorized. */
  ipfsGateway?: string;
};

function boundedBytes(parameters: {
  value: string;
  limit: number;
}): Uint8Array {
  if (parameters.value.length > parameters.limit)
    throw new Error("Metadata exceeds client size limit");
  const bytes = new TextEncoder().encode(parameters.value);
  if (bytes.length > parameters.limit)
    throw new Error("Metadata exceeds client size limit");
  return bytes;
}

export function parseContractMetadata(parameters: {
  json: string;
}): ContractMetadata {
  try {
    boundedBytes({ value: parameters.json, limit: maxMetadataBytes });
    return contractMetadataSchema.parse(JSON.parse(parameters.json));
  } catch (cause) {
    throw new Error(
      "Invalid application metadata: expected ERC-7572 JSON with nonempty name and description",
      { cause },
    );
  }
}

function decodeJsonBytes(parameters: { bytes: Uint8Array }): ContractMetadata {
  try {
    if (parameters.bytes.length > maxMetadataBytes)
      throw new Error("Metadata exceeds client size limit");
    const json = new TextDecoder("utf-8", {
      fatal: true,
      ignoreBOM: true,
    }).decode(parameters.bytes);
    return parseContractMetadata({ json });
  } catch (cause) {
    throw new Error("Invalid application metadata document", { cause });
  }
}

function decodeDataUri(parameters: { uri: string }): ContractMetadata {
  try {
    const match =
      /^data:application\/json(?:(;utf8|;charset=utf-8))?(;base64)?,([\s\S]*)$/i.exec(
        parameters.uri,
      );
    if (!match) throw new Error("Unsupported JSON data URI encoding");
    const payload = decodeURIComponent(match[3]!);
    if (!match[2]) return parseContractMetadata({ json: payload });
    if (
      !/^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/.test(
        payload,
      )
    )
      throw new Error("Invalid Base64");
    const binary = atob(payload);
    if (btoa(binary) !== payload) throw new Error("Noncanonical Base64");
    return decodeJsonBytes({
      bytes: Uint8Array.from(binary, (character) => character.charCodeAt(0)),
    });
  } catch (cause) {
    throw new Error("Invalid application metadata data URI", { cause });
  }
}

function remoteUrl(parameters: { uri: string; ipfsGateway?: string }): string {
  if (parameters.uri.startsWith("https://")) return parameters.uri;
  if (!parameters.uri.startsWith("ipfs://"))
    throw new Error("Metadata resolution failed: unsupported URI scheme");
  if (!parameters.ipfsGateway)
    throw new Error(
      "Metadata resolution failed: IPFS requires a configured HTTPS gateway",
    );
  const gateway = new URL(parameters.ipfsGateway);
  if (
    gateway.protocol !== "https:" ||
    gateway.username ||
    gateway.password ||
    gateway.search ||
    gateway.hash ||
    gateway.pathname !== "/"
  )
    throw new Error(
      "Metadata resolution failed: IPFS gateway must be an HTTPS origin",
    );
  const match = /^ipfs:\/\/([a-zA-Z0-9]+)((?:\/[^?#]*)?)$/.exec(parameters.uri);
  if (!match) throw new Error("Metadata resolution failed: invalid IPFS URI");
  const segments = match[2]!
    .split("/")
    .slice(1)
    .map((segment) => {
      const decoded = decodeURIComponent(segment);
      if (decoded === "." || decoded === ".." || /[\\/]/.test(decoded))
        throw new Error("Metadata resolution failed: invalid IPFS path");
      return encodeURIComponent(decoded);
    });
  return `${gateway.origin}/ipfs/${match[1]}${segments.length ? `/${segments.join("/")}` : ""}`;
}

export async function resolveContractMetadata(
  parameters: MetadataResolutionOptions & { uri: string },
): Promise<ResolvedContractMetadata> {
  boundedBytes({ value: parameters.uri, limit: maxMetadataUriBytes });
  if (!parameters.uri)
    throw new Error("Invalid application metadata: contractURI() is empty");
  if (/^data:/i.test(parameters.uri))
    return {
      uri: parameters.uri,
      metadata: decodeDataUri({ uri: parameters.uri }),
    };

  let body: Uint8Array;
  try {
    const url = remoteUrl(parameters);
    const request: HttpRequest = {
      url,
      method: "GET",
      headers: [{ name: "Accept", value: "application/json" }],
      body: "0x",
      requirements: [],
    };
    validateRequest(request);
    if (!parameters.authorizeRequest)
      throw new Error("Remote metadata requires network authorization");
    await parameters.authorizeRequest({
      request,
      completedRequest: request,
      requirements: [],
    });
    const response = await executeHttpRequest({
      request,
      maxResponseBytes: maxMetadataBytes,
      fetch: (input, init) =>
        (parameters.fetch ?? globalThis.fetch)(input, {
          ...init,
          credentials: "omit",
          signal: AbortSignal.timeout(30_000),
        }),
    });
    if (response.status < 200 || response.status >= 300)
      throw new Error(`Metadata HTTP status ${response.status}`);
    body = hexToBytes(response.body);
  } catch (cause) {
    throw new Error(
      `Metadata resolution failed: ${cause instanceof Error ? cause.message : String(cause)}`,
      { cause },
    );
  }
  return { uri: parameters.uri, metadata: decodeJsonBytes({ bytes: body }) };
}

export async function readContractMetadata(
  parameters: MetadataResolutionOptions & {
    address: Address;
    ethCall: EthCall;
  },
): Promise<ResolvedContractMetadata> {
  let result;
  try {
    result = await parameters.ethCall({
      to: parameters.address,
      data: encodeFunctionData({
        abi: contractMetadataAbi,
        functionName: "contractURI",
      }),
    });
  } catch (cause) {
    throw new Error(
      "Required contractURI() call failed; application metadata could not be read",
      { cause },
    );
  }
  let uri: string;
  try {
    uri = decodeFunctionResult({
      abi: contractMetadataAbi,
      functionName: "contractURI",
      data: result,
    });
  } catch (cause) {
    throw new Error(
      "Invalid contractURI() return: required application metadata is missing or malformed",
      { cause },
    );
  }
  return resolveContractMetadata({ ...parameters, uri });
}
